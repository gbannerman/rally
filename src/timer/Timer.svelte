<script lang="ts">
  import {
    convertMillisecondsToTimerValue,
    convertTimerValueToMilliseconds,
    CountdownTimer,
  } from "./timer";
  import StartStopButton from "./StartStopButton.svelte";
  import TimerDisplay from "./TimerDisplay.svelte";
  import TimerInput from "./TimerInput.svelte";

  let timerValue = "";
  let remainingMilliseconds = 0;
  let running = false;

  const timer = new CountdownTimer(
    50,
    () => {
      running = false;
      console.log("finished");
    },
    () => {
      remainingMilliseconds = timer.timeRemaining;
    },
    () => console.error("errorrrr")
  );

  const toggle = async () => {
    if (running) {
      timer.stop();
      timerValue = convertMillisecondsToTimerValue(remainingMilliseconds);
    } else {
      const duration = convertTimerValueToMilliseconds(timerValue);
      remainingMilliseconds = duration;
      timer.start(duration);
    }
    running = !running;
  };
</script>

<div class="Timer">
  {#if running}
    <TimerDisplay {remainingMilliseconds} />
  {:else}
    <TimerInput bind:value={timerValue} />
  {/if}
  <StartStopButton {running} onToggle={toggle} />
</div>

<style>
  .Timer {
    position: absolute;
    top: 20px;
    left: calc(50% - 50px);
    width: 100px;
    display: flex;
    flex-direction: column;
    align-items: center;
  }
</style>
