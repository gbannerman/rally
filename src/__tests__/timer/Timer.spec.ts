import { act, fireEvent, render } from "@testing-library/svelte";
import userEvent from "@testing-library/user-event";
import Timer from "../../timer/Timer.svelte";

jest.useFakeTimers();

describe("Timer component", () => {
  it("renders correctly", () => {
    const { container } = render(Timer);

    expect(container).toMatchSnapshot();
  });

  it("starts off at zero", () => {
    const { getByText } = render(Timer);

    const timerInput = getByText("00:00");

    expect(timerInput).toBeVisible();
  });

  it("hides input when running", async () => {
    const { getByRole, queryByRole } = render(Timer);

    const input = getByRole("textbox");

    userEvent.type(input, "500");

    const startButton = getByRole("button");
    await fireEvent.click(startButton);

    expect(queryByRole(input)).not.toBeInTheDocument();
  });

  it("start button starts countdown timer", async () => {
    const { getByText, getByRole } = render(Timer);

    const input = getByRole("textbox");

    userEvent.type(input, "500");

    const startButton = getByRole("button");

    await fireEvent.click(startButton);

    await act(() => {
      jest.advanceTimersByTime(9_000);
    });

    expect(getByText("04:51")).toBeVisible();
  });

  it("stop button sets input to remaining time", async () => {
    const { getByRole } = render(Timer);

    let input = getByRole("textbox");

    userEvent.type(input, "500");

    const startButton = getByRole("button");
    await fireEvent.click(startButton);

    await act(async () => {
      jest.advanceTimersByTime(12_000);
    });

    const stopButton = getByRole("button");
    await fireEvent.click(stopButton);

    input = getByRole("textbox");

    expect(input).toHaveValue("04:48");
  });

  it("returns to input state when timer finished", async () => {
    const { getByRole } = render(Timer);

    let input = getByRole("textbox");

    userEvent.type(input, "5");

    let startButton = getByRole("button");
    await fireEvent.click(startButton);

    await act(async () => {
      jest.advanceTimersByTime(6_000);
    });

    startButton = getByRole("button");
    input = getByRole("textbox");

    expect(startButton).toHaveClass("start");
    expect(input).toBeVisible();
  });
});
