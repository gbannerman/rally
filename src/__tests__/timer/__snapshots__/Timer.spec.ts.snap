// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer component renders correctly 1`] = `
<body>
  <div>
    <div
      class="Timer svelte-u8xogd"
    >
      <div
        aria-label="Timer"
        class="TimerInput svelte-jd5xj"
      >
        <div
          class="timer placeholder svelte-jd5xj"
        >
          00:00
        </div>
         
        <form>
          <input
            class="timer input svelte-jd5xj"
            inputmode="numeric"
            maxlength="5"
            style="width: 10px; margin-right: 2px; padding-left: 0px; right: 0px;"
            type="text"
          />
        </form>
      </div>
       
      <button
        class="icon start svelte-znyfjv"
      >
        <svg
          class="feather feather-play "
          fill="none"
          height="100%"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="5 3 19 12 5 21 5 3"
          />
        </svg>
      </button>
    </div>
  </div>
</body>
`;
